src/Sylius/Bundle/CoreBundle/Routing/Generator/ChannelProductUrlGenerator.php

```diff
@@ -0,0 +1,105 @@
+use Sylius\Component\Core\Model\ProductInterface;
+use Sylius\Component\Core\Model\ProductTranslationInterface;
+use Sylius\Component\Locale\Context\LocaleContextInterface;
+use Sylius\Component\Locale\Model\LocaleInterface;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+
+final class ChannelProductUrlGenerator implements ChannelProductUrlGeneratorInterface
+{
+    public function __construct (
+        private LocaleContextInterface $localeContext,
+        private UrlGeneratorInterface $urlGenerator,
+        private bool $unsecuredUrls,
+    ) {
+    }
+
+    public function generate(ProductInterface $product, ChannelInterface $channel): ?string
+    {
+        /** @var Collection<array-key, ProductTranslationInterface> $productTranslations */
+        $productTranslations = $product->getTranslations();
+
+        $administratorLocaleCode = $this->localeContext->getLocaleCode();
+        $productTranslation = $this->findTranslationWithSlugForLocales($productTranslations, [$administratorLocaleCode]);
+
+        if (false !== $productTranslation) {
+            return $this->generateProductUrl($productTranslation, $channel);
+        }
+
+        /** @var string $channelDefaultLocaleCode */
+        $channelDefaultLocaleCode = $channel->getDefaultLocale()->getCode();
+        $productTranslation = $this->findTranslationWithSlugForLocales($productTranslations, [$channelDefaultLocaleCode]);
+
+        if (false !== $productTranslation) {
+            return $this->generateProductUrl($productTranslation, $channel);
+        }
+
+        $localesEnabledInChannel = $this->getLocalesCodesEnabledInChannel($channel);
+        $productTranslation = $this->findTranslationWithSlugForLocales($productTranslations, $localesEnabledInChannel);
+
+        if (false !== $productTranslation) {
+            return $this->generateProductUrl($productTranslation, $channel);
+        }
+
+        return null;
+    }
+
+    /**
+     * @param Collection<array-key, ProductTranslationInterface> $productTranslations
+     * @param array<string> $localeCodes
+     */
+    private function findTranslationWithSlugForLocales(Collection $productTranslations, array $localeCodes): ProductTranslationInterface|false
+    {
+        return $productTranslations->filter(function (ProductTranslationInterface $productTranslation) use ($localeCodes): bool {
+            return in_array($productTranslation->getLocale(), $localeCodes) &&
+                '' !== $productTranslation->getSlug() &&
+                null !== $productTranslation->getSlug()
+            ;
+        })->first();
+    }
+
+    private function generateProductUrl(ProductTranslationInterface $productTranslation, ChannelInterface $channel): string
+    {
+        $productPath = $this->urlGenerator->generate('sylius_shop_product_show', [
+            'slug' => $productTranslation->getSlug(),
+            '_locale' => $productTranslation->getLocale(),
+        ]);
```

Comment:
I have some doubts about hardcoding the route here, but it probably doesn't make sense to extract it into the service arguments, for example. 
Alternatively, you could extract a separate service that just finds the appropriate translation, but generating the route based on that could be in another service or even in the template, as it will be only getting slug and locale from that object.

Reply:
I'm not sure about that. At the end of the day, somebody has to override it somehow, but it doesn't look like a thing to be overridden.